flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/StreamTwoInputProcessor.java

```diff
@@ -184,10 +184,6 @@ public boolean processInput() throws Exception {
 		}
 		checkFinished(status, lastReadInputIndex);
 
-		if (status != InputStatus.MORE_AVAILABLE) {
```

Comment:
Some side effect of removing this code is that now `inputSelectionHandler` will be updated not immediately during this `#processInput(...)` invocation, but only during the next one, right? Could this cause any (performance?) issues?

Reply:
Yes, I refactored this part in a separate hotfix commit for integrating the logics of `inputSelectionHandler#setAvailableIndex` and `setUnavailableIndex` together before `emitNext` call. In the past these two logics are scattered before and after `emitNext` separately.

I do not think it would impact the performance, and at-least in my benchmark results it has not caused.  I am curious that are there any purpose to update the selection immediately before?

Reply:
> are there any purpose to update the selection immediately before?

I'm not sure if there is any difference now, with the updated return status code. Previously it was returning boolean "more available" for hot looping. Without up to date input availability, it could attempt to read from incorrect (empty) input and that would trigger a `volatile` `isAvailable()` access.

Reply:
Let me summary the changes a bit. There are mainly three changes:

1. Migrate `inputSelectionHandler.setUnavailableInput` to integrate with `selectNextReadingInputIndex#updateAvailability`. I think it has no behavior different with before. Just to put similar logics together for better tracing.

2. Return status instead of boolean for `StreamInputProcessor#processInput`. It is for replacing the `isFinished` method and keeping the consistent form with `emitNext` return type. It also has no behavior change, and only `StreamTask#processInput` adds two additional conditions check instead of previous `while` way. But we can also use `while` way or add the hot path to avoid performance concern.
```
if (status == InputStatus.MORE_AVAILABLE) {
     return;
}
```

3. The value returned by `StreamTwoInputProcessor#processInput` is different from before which might cause different behaviors. In the past it might return invalid status after current call, and then in the next call it might return -1 while `selectNextReadingInputIndex()` to trigger `isAvailable()` access. But now it would always return valid status by judging the next selected index with its status. So if it is not available for next selected input, it would trigger `isAvailable()` access in current call directly, no need to trigger it in next call. From this point, my previous thought was that it might save some unnecessary cost from mailbox schedule aspect.

Anyway, it is hard to verify the above points because of unstable benchmark. So we can only analyze the possible impacts in theory.